eclair-core/src/main/scala/fr/acinq/eclair/wire/Onion.scala

```diff
@@ -18,48 +18,160 @@ package fr.acinq.eclair.wire
+    varintoverflow.decode(bits).map(d => DecodeResult(d.value + (bits.length - d.remainder.length) / 8, d.remainder)))
+
+  private val amountToForward: Codec[AmountToForward] = ("amount_msat" | tu64overflow).xmap(amountMsat => AmountToForward(MilliSatoshi(amountMsat)), (a: AmountToForward) => a.amount.toLong)
+
+  private val outgoingCltv: Codec[OutgoingCltv] = ("cltv" | tu32).xmap(cltv => OutgoingCltv(CltvExpiry(cltv)), (c: OutgoingCltv) => c.cltv.toLong)
+
+  private val outgoingChannelId: Codec[OutgoingChannelId] = (("length" | constant(hex"08")) :: ("short_channel_id" | shortchannelid)).as[OutgoingChannelId]
+
+  private val onionTlvCodec = discriminated[OnionTlv].by(varint)
+    .typecase(UInt64(2), amountToForward)
+    .typecase(UInt64(4), outgoingCltv)
+    .typecase(UInt64(6), outgoingChannelId)
+
+  val tlvPerHopPayloadCodec: Codec[TlvStream[OnionTlv]] = TlvCodecs.lengthPrefixedTlvStream[OnionTlv](onionTlvCodec).complete
+
+  private val legacyRelayPerHopPayloadCodec: Codec[RelayLegacyPayload] = (
+    ("realm" | constant(ByteVector.fromByte(0))) ::
+      ("short_channel_id" | shortchannelid) ::
+      ("amt_to_forward" | millisatoshi) ::
+      ("outgoing_cltv_value" | cltvExpiry) ::
+      ("unused_with_v0_version_on_header" | ignore(8 * 12))).as[RelayLegacyPayload]
+
+  private val legacyFinalPerHopPayloadCodec: Codec[FinalLegacyPayload] = (
+    ("realm" | constant(ByteVector.fromByte(0))) ::
+      ("short_channel_id" | ignore(8 * 8)) ::
+      ("amount" | millisatoshi) ::
+      ("expiry" | cltvExpiry) ::
+      ("unused_with_v0_version_on_header" | ignore(8 * 12))).as[FinalLegacyPayload]
+
+  case class MissingRequiredTlv(tag: UInt64) extends Err {
+    // @formatter:off
+    val failureMessage: FailureMessage = InvalidOnionPayload(tag, 0)
+    override def message = failureMessage.message
+    override def context: List[String] = Nil
+    override def pushContext(ctx: String): Err = this
+    // @formatter:on
+  }
+
+  val relayPerHopPayloadCodec: Codec[RelayPayload] = fallback(tlvPerHopPayloadCodec, legacyRelayPerHopPayloadCodec).narrow({
+    case Left(tlvs) if tlvs.get[AmountToForward].isEmpty => Attempt.failure(MissingRequiredTlv(UInt64(2)))
+    case Left(tlvs) if tlvs.get[OutgoingCltv].isEmpty => Attempt.failure(MissingRequiredTlv(UInt64(4)))
+    case Left(tlvs) if tlvs.get[OutgoingChannelId].isEmpty => Attempt.failure(MissingRequiredTlv(UInt64(6)))
+    case Left(tlvs) => Attempt.successful(RelayTlvPayload(tlvs))
+    case Right(legacy) => Attempt.successful(legacy)
+  }, {
+    case legacy: RelayLegacyPayload => Right(legacy)
+    case RelayTlvPayload(tlvs) => Left(tlvs)
+  })
+
+  val finalPerHopPayloadCodec: Codec[FinalPayload] = fallback(tlvPerHopPayloadCodec, legacyFinalPerHopPayloadCodec).narrow({
+    case Left(tlvs) if tlvs.get[AmountToForward].isEmpty => Attempt.failure(MissingRequiredTlv(UInt64(2)))
+    case Left(tlvs) if tlvs.get[OutgoingCltv].isEmpty => Attempt.failure(MissingRequiredTlv(UInt64(4)))
```

Comment:
Yes but it's a sender restriction. I think it's harmless as a receiver to accept and ignore it instead of failing the payment, that means we would be rejecting payments that are otherwise valid and are giving us money :D

Reply:
Fair enough